### ISSUE TITLE
To-Do Updates to Parsons/Order Blocks Element

### ISSUE THREAD

**Issue by nathanielweinman** [PrairieLearn/PrairieLearn#3476]
Below are some nice-to-haves for the Parsons/Order Blocks element once it's in! https://github.com/PrairieLearn/PrairieLearn/pull/2760

RELATED: the order blocks saga has spawned another issue -  #4652, regarding the modification of the partial credit grading algorithm 

- [x] Remove `inline-block` from the correct answer and submitted answer formatting, so each block will be on its own line and the answers will be "vertically" displayed.
- [x] When using vertical layout, set `width=100%`.
- [x] In the CSS, prefix all classes (and IDs) with `pl-order-blocks-` to avoid conflicts with other elements and user-defined styles.
- [x] Add help text in the UI depending on attributes. For example, specify whether the order of the solution blocks matters, and provide instructions on indentation if it's enabled. Think about every attribute and whether it needs help text or instructions in the UI.
- [x] add some some sort of "help" (?) button that once hovered over, gives information about the expected format for the answer
- [x] Support more than 4 indents (or at least make it configurable)
- [x] Update long text to wrap-around instead of ellipses: (e.g., https://github.com/ace-lab/berk-parsons/blob/master/app/static/vendors/parsons/parsons.css#L34-L47)
- [x] Combine `external-grader` and `permutation-mode` options (possibly controversial)
- [x] ~~Add attribute `min-correct` and `max-correct`, similar to `min-incorrect` and `max-incorrect`~~ Need more input for this one, as existing attribute behaviours might be affected 
- [x] if possible, improve the current method of parsing html, which uses xpath and is overkill
- [x] **Use Bootstrap for styling**
- [x] Consider adding an `inline` option that would display all the blocks as inline (in the source area, ~~the solution area, the correct answer, and submitted answers~~). Only source area was added, as inline drag and drop detection is extremely buggy
- [x] add a help text when `indentation` is set to true to make it obvious to students that they should indent their answers
- [x] **Add support for using DAG to grade proofs**
- [ ] Add an option to alphabetized lines in `shuffle-options` (motivated by https://dl.acm.org/doi/pdf/10.1145/3017680.3017704). It makes the lines more easily findable for students but likely doesn't give anything away compared to pure shuffling.
- [ ] Give partial credit when using `grading-method="ordered"` or `"ranking"`. For example, for `ordered` we could use `1 - max(0, lev(correct_answer, submitted_answer)/len(correct_answer))`, where `lev()` is the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance). Here we would treat each block as a unique symbol (for "ordered") or blocks of the same ranking as the same symbol (for "ranking"). See http://ceur-ws.org/Vol-2734/paper11.pdf (note from @SethPoulsen: I'm not really a fan of this idea, happy to discuss why. Of course, we could have the method of partial credit calculation be configurable and then people can do what they want).
- [ ] Add optional hints to students for incorrect answers. For example, we could say "there are XX blocks still in the source area that should be in the solution, and you have YY extra blocks in the solution that do not belong there". Similarly, we could provide hints when the set of solution blocks is correct, but the order is wrong. For example, "Block XXX is in the wrong position in the solution".
- [x] Support `<pl-figure>` and other elements inside `<pl-answer>` inside `<pl-order-blocks>`
- [ ] show correct indentation when showing the correct answer
- [ ] BUG: UI allows you to place a block with negative indentation
- [ ] BUG: the element sometimes show the partial credit you _would have_ gotten if partial credit was enabled, even when its not
- [ ] Add vertical lines to make the indenting affordance clear (e.g., https://github.com/ace-lab/berk-parsons/blob/79cd59dca70e0ac8156c9386bc8c0b1fcbd94fc0/app/static/vendors/parsons/parsons.js#L1470-L1477, https://github.com/RunestoneInteractive/RunestoneComponents/blob/05d5a5d0a03369fb7726474d68f07b7dbd2d7c26/runestone/parsons/css/parsons.css#L100-L109)
- [ ] use `em` rather than pixel to set indent locations, so that the indent can be exactly 4 characters
- [x] BUG: when multiple `pl-order-blocks` are in a single question, having a blank answer for any of them results in "invalid submission" for all of them

cc @mfsilva22, sorry for the delay, but moved my future PR comments here.

**Comment by mwest1066** [comment]
I've been editing the main comment to add more items.

**Comment by mwest1066** [comment]
@Frozemint: I think you might have already implemented a couple of the above items (wrapping rather than ellipses, and combining external-grader and permutation-mode)? If this is the case, could you check them off above?

**Comment by nealterrell** [comment]
Suggestion: an option to enable code styling for the blocks, instead of the default font. It would be sufficient for `<pl-code>` to work instead of a block, but even more convenient would be a language attribute for the entire element.

**Comment by jonatanschroeder** [comment]
One more issue: currently if the grading-method is "external", the answer panel assumes that the code is Python and reads the solution in `tests/ans.py`, and the `correct` attribute in `pl-answer` is not allowed. Please make this more generic

P.S.: I have random Assembly code generated in `server.py` that updates `data["correct_answers"][answers_name]`, supporting at least that would be useful. For now I'm using `pl-hide-in-panel` with custom answer panel.

**Comment by jonatanschroeder** [comment]
There is an issue as well with duplicate elements. The current implementation has some undocumented and non-obvious behaviour in case there is more than one answer with the same text. It would be useful to have an appropriate way to handle lines that can be used more than once, either by fully supporting answers with the same text, or by having some property that allows the same line to be retained on the left panel when dragged to the final answer.

**Comment by Frozemint** [comment]
Yes, there is a substantial rewrite going on at #3837, where I am adding other features from the todo list like inline CSS styling and help text, as well as a rewrite on how answers are parsed and stored (the current implementation uses a lot of arrays... the new rewrite uses dictionaries). I'll add your suggestions and to generalise the external grading method issue. I think I will use the approach where we properly handle tiles with duplicate `innerHTML`, since the other approach seems more complicated as it involves changing the dragging behaviour.

**Comment by mwest1066** [comment]
@SethPoulsen could the remaining items here be folded into your upcoming PR? Or are there things which we have decided against implementing?

**Comment by SethPoulsen** [comment]
The only thing I am doing is the DAG grader. As for the other things on this list, I don't see any reason _not_ to implement them, but no one has wanted the features badly enough to actually do it. 

I'm happy to help review PRs or give pointers to places in the code if someone wants to implement any of these though.

**Comment by mwest1066** [comment]
> The only thing I am doing is the DAG grader. As for the other things on this list, I don't see any reason _not_ to implement them, but no one has wanted the features badly enough to actually do it.
> 
> I'm happy to help review PRs or give pointers to places in the code if someone wants to implement any of these though.

Ok, sounds good.

**Comment by mwest1066** [comment]
@nathanielweinman what's your status with this element? Are you interested in extending the current `<pl-order-blocks>` element in PL? Maybe even to include faded blocks? 😄

**Comment by Frozemint** [comment]
I could work on the smaller quality of life features in the checklist.

**Comment by SethPoulsen** [comment]
I've just added "show correct indentation when showing the correct answer". 

I know @CoffeePoweredComputers has it working in a local fork, but hasn't had the time to clean it up.

**Comment by mwest1066** [comment]
> I could work on the smaller quality of life features in the checklist.

Yes please! That would be awesome!

**Comment by mwest1066** [comment]
> I've just added "show correct indentation when showing the correct answer".
> 
> I know @CoffeePoweredComputers has it working in a local fork, but hasn't had the time to clean it up.

Sweet! @CoffeePoweredComputers any chance of submitting a PR for this?

**Comment by CoffeePoweredComputers** [comment]
> Sweet! @CoffeePoweredComputers any chance of submitting a PR for this?

Yep! As Seth mentioned I still need to clean it up but once that's done I'd be happy to submit a PR.

**Comment by mwest1066** [comment]
> Yep! As Seth mentioned I still need to clean it up but once that's done I'd be happy to submit a PR.

Great, thanks!

**Comment by SethPoulsen** [comment]
There is enough going on with `pl-order-blocks` right now that I decided we should have our own project board, I made it here: `https://github.com/PrairieLearn/PrairieLearn/projects/21`. If any of the issues listed in the big list above are particularly important to you right now, feel free to split them off to their own issue and put them in the board.

I'm also happy to give people pointers to places in the code, and to review code, if anyone wants to work on any of these.
